pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/SegmentCompletionFSMFactory.java

```
@@ -0,0 +1,132 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.helix.core.realtime;
+
+import com.google.common.base.Preconditions;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.pinot.common.metadata.segment.SegmentZKMetadata;
+import org.apache.pinot.common.utils.LLCSegmentName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SegmentCompletionFSMFactory {
+  private SegmentCompletionFSMFactory() {
+    // Private constructor to prevent instantiation
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentCompletionFSMFactory.class);
+  private static final Map<String, Class<? extends SegmentCompletionFSM>> FSM_CLASS_MAP = new HashMap<>();
+
+  // Static block to register the default FSM
+  static {
+    try {
+      Class<?> clazz = Class.forName(BlockingSegmentCompletionFSM.class.getCanonicalName());
+      register(SegmentCompletionConfig.DEFAULT_FSM_SCHEME, (Class<? extends SegmentCompletionFSM>) clazz);
+      LOGGER.info("Registered default BlockingSegmentCompletionFSM");
+    } catch (Exception e) {
+      LOGGER.error("Failed to register default BlockingSegmentCompletionFSM", e);
+      throw new RuntimeException("Failed to register default BlockingSegmentCompletionFSM", e);
+    }
+  }
+
+  /**
+   * Registers an FSM class with a specific scheme/type.
+   *
+   * @param scheme The scheme or type key.
+   * @param fsmClass The class for the FSM.
+   */
+  public static void register(String scheme, Class<? extends SegmentCompletionFSM> fsmClass) {
+    Preconditions.checkNotNull(scheme, "Scheme cannot be null");
+    Preconditions.checkNotNull(fsmClass, "FSM Class cannot be null");
+    if (FSM_CLASS_MAP.containsKey(scheme)) {
+      LOGGER.warn("Overwriting existing FSM class for scheme {}", scheme);
```

Let's throw exception here to avoid unexpected registry:
`Preconditions.checkState(FSM_CLASS_MAP.put(scheme, fsmClass) == null, ...)`